### PR TITLE
state_getSize RPC for storage maps

### PR DESCRIPTION
**Pull request by kianenigma**
Currently this rpc endpoint only works for state values. 

Now, if no data exists in the given storage key, we check it is a (double) map and any other key-pairs are stored with that prefix. If so, we return the sum of all values.

### CODE REVIEW

File: client/rpc/src/state/mod.rs [hunk 1]
```diff
@@ -102,8 +102,19 @@ pub trait StateBackend<Block: BlockT, Client>: Send + Sync + 'static
 		block: Option<Block::Hash>,
 		key: StorageKey,
 	) -> FutureResult<Option<u64>> {
-		Box::new(self.storage(block, key)
-			.map(|x| x.map(|x| x.0.len() as u64)))
+		let value_len = self.storage(block, key.clone()).map(|x| x.map(|x| x.0.len() as u64));
+		let pairs_len = self.storage_pairs(block, key.clone())
```
Comment (tomusdrw) [inline client/rpc/src/state/mod.rs:106]: Shouldn't that query be done conditionally only if `value_len` results into `None`?

Comment (kianenigma) [reply client/rpc/src/state/mod.rs:106]: yeah that makes more sense and I initially started with that, but failed to get it to work. Ideally I want: only if the first one is `Ok(None)`, then try the second one. 

But I am not sure if this is even possible? I am but new to futures.

Comment (tomusdrw) [reply client/rpc/src/state/mod.rs:106]: Depends on the futures version we have here, but it seems it's 0.1 and I think it should be something like this:
```rust
value_len.and_then(|result| match result {
  None => Either::A(self.storage_pairs(...)...),
  Some(v) => Either:B(futures::done(Ok(Some(v))),
})
```

Comment (kianenigma) [reply client/rpc/src/state/mod.rs:106]: I'm pretty sure I tried something like that and failed once, but let me try again and get back.

Comment (kianenigma) [reply client/rpc/src/state/mod.rs:106]: Okay so this was the first time I was working with futures and the main point of confusion was the versioning going on (which makes me appreciate more and more the standardised async/await). I wasn't aware that the `rpc` us using and re-exporting `0.1` and the current crate is using `0.3`, and I was getting strange errors all over the place.

Comment (bkchr) [reply client/rpc/src/state/mod.rs:108]: The `storage_pairs` function will return the results directly. You need to do this differently here.

File: client/rpc/src/state/mod.rs [hunk 2]
```diff
@@ -102,8 +102,26 @@ pub trait StateBackend<Block: BlockT, Client>: Send + Sync + 'static
...
+				} else {
+					None
+				}
+			});
+
+		// if value size is there, use that, else execute another future that looks for map len.
+		let value_or_map = value_len.and_then(|result| match result {
+			Some(size) => Either::A(done(Ok(Some(size)))),
+			None => Either::B(pairs_len),
+		 });
```
Comment (bkchr) [inline client/rpc/src/state/mod.rs:122]:
```suggestion
		let pairs_len = || self.storage_pairs(block, key.clone())
			.map(|kv| {
				let item_sum = kv.iter().map(|(_, v)| v.0.len() as u64).sum::<u64>();
				if item_sum > 0 {
					Some(item_sum)
				} else {
					None
				}
			});

		// if value size is there, use that, else execute another future that looks for map len.
		let value_or_map = value_len.and_then(|result| match result {
			Some(size) => Either::A(done(Ok(Some(size)))),
			None => Either::B(pairs_len()),
		 });
```

File: client/rpc/src/state/mod.rs (hunk 1 above)
Comment (kianenigma) [reply client/rpc/src/state/mod.rs:108]: can you explain more? I don't see the difference in what you suggested.

Comment (bkchr) [reply client/rpc/src/state/mod.rs:108]: When you call a function that returns a future, you can also directly return the result. This means the first time the future is polled the result is ready. However this also means that the operation is always executed, in this case the scanning of the storage. I propose that you put the scanning call into a closure and only call this closure if the first storage call returned a none. This means you will only scan the storage when the value does not exists vs query it always.

Comment (kianenigma) [reply client/rpc/src/state/mod.rs:108]: I exactly wanted to do as you explain: only query the second if the first one is `None`. This is how I thought: calling `self.storage_pairs` and storing it in `pairs_len` executes nothing, because it is not yet polled. I thought that also afterwards when I pass it to `and_then`, it is would be only polled in the `None` arm, but apparently not.

**Comment by kianenigma** [comment]
with a gratitude to @bkchr:

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
bot merge
